File: C++/865. Smallest Subtree with all the Deepest Nodes .cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    int maxdepth = 0;
    TreeNode *res;
    int postOrder(TreeNode *root, int depth)
    {
        if (root == NULL)
            return depth;
        int leftDepth = postOrder(root->left, depth + 1);
        int rightDepth = postOrder(root->right, depth + 1);
        if (leftDepth == rightDepth && leftDepth >= maxdepth)
        {
            res = root;
            maxdepth = leftDepth;
        }
        return max(leftDepth, rightDepth);
    }
    TreeNode *subtreeWithAllDeepest(TreeNode *root)
    {
        postOrder(root, 0);
        return res;
    }
};

File: 算法/Python/865. Smallest Subtree with all the Deepest Nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        self.res = None
        self.maxdepth = 0

        def dfs(root, depth):
            if(root == None):
                return depth
            left = dfs(root.left, depth+1)
            right = dfs(root.right, depth+1)
            if(left == right and left >= self.maxdepth):
                self.maxdepth = left
                self.res = root
            return max(left, right)
        dfs(root, 0)
        return self.res
